AMI diarization and overlap detector

desh2608 opened the pull request:
This PR contains the following:
* Diarization recipe for AMI mix-headset data --- using AHC, spectral, and VBx clustering
* Training and evaluation of a TDNN-LSTM overlap detector

desh2608 commented:
@leibny @HuangZiliAndy could you take a look at this PR when you have some time?

danpovey commented:
Is it necessary to commit those rttm files?  They are quite large, I'm concerned about bulking up the Kaldi repo.

desh2608 commented:
I removed the RTTM files and added script to prepare them from annotations (similar to s5b). Also changed some files to soft links to the `s5b/local` directory.

danpovey commented:
Thanks!  will wait for the others to do initial review, but please ping if you don't hear from them within a week.

HuangZiliAndy commented:
Hi Desh, would you mind if I take a look in this week? (Get busy with schoolwork this two days)

leibny commented:
I will take a look this week as well.

Cheers,
Paola



On Sun, Sep 13, 2020 at 10:32 AM Zili Huang <notifications@github.com>
wrote:

> Hi Desh, would you mind if I take a look in this week? (Get busy with
> schoolwork this two days)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-691678946>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFSFVXUTS6SAHZ4OQXOBHDTSFTJXLANCNFSM4RHZTEUQ>
> .
>

HuangZiliAndy commented on egs/ami/s5c/local/ami_text_prep.sh:
```diff
@@ -0,0 +1 @@
+../../s5b/local/ami_text_prep.sh
```
There are some files missing. local/split_train.orig, split_dev.orig, split_eval.orig, nnet3.

HuangZiliAndy commented:
Hi Desh, I am taking a look at the pull request. It may take several days since I am trying to go through the process. I will inform you when I finish. (You don't need to take look before I finish)

HuangZiliAndy commented on egs/ami/s5c/run.sh:
```diff
@@ -0,0 +1,165 @@
+fi
+
+if [ $stage -le 4 ]; then
+  echo "$0: preparing a AMI training data to train PLDA model"
```
Perhaps here some more documentation is needed. I think you may add something about what this step is for. (feature norm)

HuangZiliAndy commented on egs/ami/s5c/run.sh:
```diff
@@ -0,0 +1,165 @@
+if [ $stage -le 5 ]; then
+  echo "$0: extracting x-vector for PLDA training data"
+  utils/fix_data_dir.sh data/plda_train
+  diarization/nnet3/xvector/extract_xvectors.sh --cmd "$train_cmd --mem 10G" \
```
I think the x-vector training part is missing. So maybe we should add some download links?

leibny replied:
It is mentioned at the beginning of the script. But perhaps it is worth a reminder here.

johnjosephmorgan commented:
Hi Desh, I cloned your pr. I found 3 scripts missing. local/split_{
dev,eval,train}.orig.

johnjosephmorgan commented:
Hi Desh,  I got this pandas error when running stage 2:
./run.sh: preparing train set..
read annotations to get segments
read audios
Traceback (most recent call last):
  File "local/prepare_data.py", line 117, in <module>
    make_diar_data(**vars(args))
  File "local/prepare_data.py", line 95, in make_diar_data
    df_wav = find_audios(wav_path, file_list)
  File "local/prepare_data.py", line 56, in find_audios
    df_wav = df_wav.loc[df_wav['key'].str.contains(file_names_str)].sort_values('key')
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/generic.py", line 4372, in __getattr__
    return object.__getattribute__(self, name)
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/accessor.py", line 133, in __get__
    accessor_obj = self._accessor(obj)
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/strings.py", line 1895, in __init__
    self._validate(data)
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/strings.py", line 1917, in _validate
    raise AttributeError("Can only use .str accessor with string "
AttributeError: Can only use .str accessor with string values, which use np.object_ dtype in pandas

johnjosephmorgan commented:
The problem seems to be with the file_names_str variable. Here is what that variable contains for me:
ES2016a|ES2007d|ES2008d|IN1005|TS3010d|EN2001a|ES2003c|ES2015a|IS1007c|ES2008a|TS3006d|TS3005b|EN2006b|ES2008b|IN1008|IN1016|ES2006b|ES2006c|ES2002b|ES2003d|TS3008b|EN2001e|ES2009a|IS1006c|IS1005a|ES2007c|ES2006a|EN2005a|TS3011b|IS1006d|EN2006a|TS3006c|TS3012c|EN2009d|IS1000b|IS1005c|IS1007a|ES2010d|ES2012a|ES2012c|TS3009c|TS3007a|ES2009b|IN1007|ES2007b|EN2009c|IS1006a|IN1012|TS3009a|TS3011a|TS3005d|ES2016c|ES2016b|IS1002c|TS3009b|IS1001d|IS1002d|EN2001b|TS3011c|ES2013c|ES2005a|ES2010b|ES2002d|ES2007a|ES2014a|ES2009d|ES2005c|IS1004c|ES2015d|ES2003a|IS1002b|EN2004a|IB4005|ES2015b|IN1001|ES2006d|TS3011d|TS3007b|IS1000d|TS3008a|IS1004d|IS1006b|ES2010c|ES2014c|TS3006a|ES2013b|ES2010a|IS1005b|ES2013a|EN2009b|TS3008d|TS3010b|ES2002a|ES2014d|TS3012d|TS3007d|ES2015c|IN1009|IS1003d|ES2003b|TS3008c|TS3012a|TS3007c|TS3009d|ES2009c|ES2005b|TS3010a|IS1003b|IS1003c|ES2012b|ES2002c|IS1004a|IS1000c|ES2014b|IS1001c|IS1001a|IS1007b|ES2013d|EN2003a|TS3012b|IN1002|IN1013|IS1007d|EN2001d|IN1014|ES2012d|IS1004b|ES2005d|IS1001b|TS3005c|TS3006b|TS3010c|TS3005a|IS1003a|ES2016d|ES2008c|IS1000a

johnjosephmorgan commented:
Hi Desh, In local/prepare_data.py there is a find command that searches for files with  Mixed-Headset.wav in their names. I can't find any such files in my version of the AMI corpus. Were the file names modified. Can you give me a path to one of these files?

desh2608 commented:
> Hi Desh, In local/prepare_data.py there is a find command that searches for files with Mixed-Headset.wav in their names. I can't find any such files in my version of the AMI corpus. Were the file names modified. Can you give me a path to one of these files?

Hi @johnjosephmorgan , they are in paths like: `amicorpus/EN2001a/audio/EN2001a.Mix-Headset.wav`

johnjosephmorgan commented:
Thanks Desh, I got the data. Now I am trying to run step 4. The local/nnet3 directory is missing. jjm

desh2608 commented:
@johnjosephmorgan thanks for pointing out. I have added them now.

johnjosephmorgan commented:
@desh2608 I ran into a problem in stage 4 in the xvector feature prep. The utils/fix_data_dir.sh in stage 5 fails on the data/plda_train directory. Stage 4 runs without errors. I think you need to copy the data/train/segments file to the data/plda_train directory.

desh2608 commented on egs/ami/s5c/local/nnet3/xvector/prepare_feats.sh:
```diff
@@ -0,0 +1,89 @@
+cp $data_in/utt2spk $data_out/utt2spk
+cp $data_in/spk2utt $data_out/spk2utt
+cp $data_in/wav.scp $data_out/wav.scp
+for f in $data_in/segments $data_in/segments/vad.scp ; do
```
@johnjosephmorgan the `segments` file is copied to plda_train here. I will run the stages again to see why it must be failing.

johnjosephmorgan commented:
OK @desh2608 my mistake. I had copied a prepare_feats.sh from another recipe before you put the correct one there. Sorry about that.

johnjosephmorgan commented:
@desh2608 I have previously run the Voxceleb V2 recipe. That recipe uses MFCC features of dimension 30. Your recipe uses MFCC feature vectors of dimension 40.  I am continuing the recipe with 30 dimensional vectors. Do you think this will be a problem? I reran from stage 3 with the modified config file.

desh2608 commented:
@johnjosephmorgan for the CHiME-6 recipe, we used 40-dim MFCCs so that the same features could be used for x-vector extractor, SAD, as well as the acoustic model. It shouldn't make much of a difference IMO.

johnjosephmorgan commented:
Hi @desh2608 I am trying to run stage 8. The overlap training uses scripts in the steps/overlap directory. That directory is in your fork but not in the kaldi master. For now I'll copy the directory from your fork to my kaldi installation, but I thought I'd let you know that the steps/overlap directory is not available in Kaldi yet.

desh2608 commented:
@johnjosephmorgan yes I created that directory since I figured other recipes may want to use it, similar to the `steps/segmentation` directory

johnjosephmorgan commented:
@desh2608 The following script is missing:
steps/overlap/train_tdnn_1a.sh

RuslanSel commented:
Hi @desh2608. The local/overlap/prepare_overlap_graph.py from local/detect_overlaps.sh is missing.

johnjosephmorgan commented:
Hi @desh2608 Don't you think the file:
steps/overlap/train_tdnn_lstm_1a.sh 
should go in a tuning directory?

Something like:
local/overlap/tuning/train_tdnn_lstm_1a.sh

desh2608 commented:
@RuslanSel thanks for pointing out. 
@johnjosephmorgan yeah you are right. I have fixed these issues now. Hopefully should work.

johnjosephmorgan commented:
@desh2608
I see the changes in local/train_overlap_detector.sh and the deletion in
wsj/s5/steps/overlap
but I do not see
local/overlap


> On Sep 28, 2020, at 4:33 PM, Desh Raj <notifications@github.com> wrote:
> 
> 
> @RuslanSel <https://github.com/RuslanSel> thanks for pointing out.
> @johnjosephmorgan <https://github.com/johnjosephmorgan> yeah you are right. I have fixed these issues now. Hopefully should work.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-700266079>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJL7AG5Q6EIMBZOHOTYPH3SIDXKJANCNFSM4RHZTEUQ>.
>

desh2608 commented:
@johnjosephmorgan sorry it seems I have been distracted today! Pushed the `local/overlap` directory now

RuslanSel commented:
@desh2608. Looks like local/overlap/prepare_overlap_graph.py still missing.

johnjosephmorgan commented:
Hi @desh2608 I am getting an extra affix _1a appended to a directory name.
I think the problem is in
local/train_overlap_detector.sh
where 
local/overlap/run_tdnn_lstm_1a.sh
gets called.
The argument to the --dir option  has an affix _1a appended to it.
I think in 
local/overlap/run_tdnn_lstm_1a.sh
another affix _1a gets appended  to the directory name.

johnjosephmorgan commented:
desh2608
In the script
steps/overlap/post_process_output.sh
on line 59, the script
local/overlap/output_to_rttm.py
is called.
This needs to be fixed.
I'm not sure if you just need to change the local to steps or if you want to put this script under local/overlap.

johnjosephmorgan commented:
@desh2608
BTW: I changed the local to steps and stage 9 finished.
I ran through the whole recipe.

desh2608 commented:
@johnjosephmorgan thanks, I'll change it. I think it's better to keep it under steps since it is a generic script.

desh2608 commented:
@danpovey perhaps this can be merged.

pzelasko commented:
Out of curiosity, what DERs did you achieve with these models? I think it would be useful to add a RESULTS file for this directory with this information (also for others to know whether they successfully reproduced the models).

johnjosephmorgan commented:
FWIW, here is the output of running stage 9 where the evaluation takes
place. I don't understand the results yet.
./run.sh: performing overlap detection on dev
--convert_data_dir_to_whole true --output-scale 1 2 1 data/dev
exp/overlap_1a/tdnn_lstm_1a_1a exp/overlap_1a/dev
diff: exp/overlap_1a/dev/final.raw: No such file or directory
steps/nnet3/compute_output.sh --nj 15 --cmd run.pl --iter final
--extra-left-context 0 --extra-right-context 0
--extra-left-context-initial -1 --extra-right-context-final -1
--frames-per-chunk 300 --apply-exp true --frame-subsampling-factor 1
data/dev_whole exp/overlap_1a/dev/overlap exp/overlap_1a/dev
utils/data/get_utt2dur.sh: data/dev_whole/utt2dur already exists with
the expected length.  We won't recompute it.
local/detect_overlaps.sh: Decoding output
utils/data/get_utt2dur.sh: data/dev_whole/utt2dur already exists with
the expected length.  We won't recompute it.
local/detect_overlaps.sh: Created output overlap RTTM at
exp/overlap_1a/dev/rttm_overlap
./run.sh: evaluating output..
MISSED SPEAKER TIME =   1711.01 secs ( 40.9 percent of scored speaker time)
FALARM SPEAKER TIME =   1062.66 secs ( 25.4 percent of scored speaker time)
./run.sh: performing overlap detection on eval
--convert_data_dir_to_whole true --output-scale 1 2 1 data/eval
exp/overlap_1a/tdnn_lstm_1a_1a exp/overlap_1a/eval
utils/fix_data_dir.sh: file data/eval_whole/utt2spk is not in sorted
order or not unique, sorting it
fix_data_dir.sh: kept all 16 utterances.
fix_data_dir.sh: old files are kept in data/eval_whole/.backup
fix_data_dir.sh: kept all 16 utterances.
fix_data_dir.sh: old files are kept in data/eval_whole/.backup
steps/make_mfcc.sh --mfcc-config conf/mfcc_hires.conf --nj 16 --cmd
run.pl --write-utt2num-frames true data/eval_whole
utils/validate_data_dir.sh: Successfully validated data-directory
data/eval_whole
steps/make_mfcc.sh: [info]: no segments file exists: assuming wav.scp
indexed by utterance.
steps/make_mfcc.sh: Succeeded creating MFCC features for eval_whole
steps/compute_cmvn_stats.sh data/eval_whole
Succeeded creating CMVN stats for eval_whole
fix_data_dir.sh: kept all 16 utterances.
fix_data_dir.sh: old files are kept in data/eval_whole/.backup
diff: exp/overlap_1a/eval/final.raw: No such file or directory
steps/nnet3/compute_output.sh --nj 16 --cmd run.pl --iter final
--extra-left-context 0 --extra-right-context 0
--extra-left-context-initial -1 --extra-right-context-final -1
--frames-per-chunk 300 --apply-exp true --frame-subsampling-factor 1
data/eval_whole exp/overlap_1a/eval/overlap exp/overlap_1a/eval
utils/data/get_utt2dur.sh: data/eval_whole/utt2dur already exists with
the expected length.  We won't recompute it.
local/detect_overlaps.sh: Decoding output
utils/data/get_utt2dur.sh: data/eval_whole/utt2dur already exists with
the expected length.  We won't recompute it.
local/detect_overlaps.sh: Created output overlap RTTM at
exp/overlap_1a/eval/rttm_overlap
./run.sh: evaluating output..
MISSED SPEAKER TIME =   1868.99 secs ( 46.1 percent of scored speaker time)
FALARM SPEAKER TIME =    856.43 secs ( 21.1 percent of scored speaker time)


On 10/3/20, Piotr Żelasko <notifications@github.com> wrote:
> Out of curiosity, what DERs did you achieve with these models? I think it
> would be useful to add a RESULTS file for this directory with this
> information (also for others to know whether they successfully reproduced
> the models).
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-703187129

desh2608 commented:
@johnjosephmorgan those are somewhat worse than what I got (36.5% missed and 16.1% false alarm). Perhaps there is some hyperparameter config that is different. I'll check and confirm.

@pzelasko The DERs on dev/eval are:
AHC: 27.0/28.3
Spectral: 27.6/26.9
VBx: 26.8/26.2
Note that these are with oracle SAD.

johnjosephmorgan commented:
@desh2806 recall that I trained the Voxceleb models on MFCC vectors of dimension 30.
Could this explain the worse results?


> On Oct 4, 2020, at 12:18 PM, Desh Raj <notifications@github.com> wrote:
> 
> 
> @johnjosephmorgan <https://github.com/johnjosephmorgan> those are somewhat worse than what I got (36.5% missed and 16.1% false alarm). Perhaps there is some hyperparameter config that is different. I'll check and confirm.
> 
> @pzelasko <https://github.com/pzelasko> The DERs on dev/eval are:
> AHC: 27.0/28.3
> Spectral: 27.6/26.9
> VBx: 26.8/26.2
> Note that these are with oracle SAD.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-703278954>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJL7ABLI252O2YJVBKAJ2LSJCN5XANCNFSM4RHZTEUQ>.
>

desh2608 commented:
@johnjosephmorgan that shouldn't affect the overlap detection result. The overlap detector is trained separately in [this](https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5c/local/train_overlap_detector.sh) script.

johnjosephmorgan commented:
@desh2608
I have a different question:
Could you train a speech activity detector with the same AMI corpus you used to train the overlap detector?
I ask because I'd like to use 16k data to train an SAD instead of the 8k data used in the aspire recipe.

On Oct 4, 2020, at 3:23 PM, Desh Raj <notifications@github.com> wrote:OK 
> 
> 
> @johnjosephmorgan <https://github.com/johnjosephmorgan> that shouldn't affect the overlap detection result. The overlap detector is trained separately in this <https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5c/local/train_overlap_detector.sh> script.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-703303243>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJL7AAWAJQ4MT46DXCH6YDSJDDUHANCNFSM4RHZTEUQ>.
>

johnjosephmorgan commented:
@desh2608 In local/train_overlap_detector.sh you say that at decode time the overlap detector could be used to do SAD. Is it a matter of consolidating the single and overlap labels as speech?

desh2608 commented:
Yes it can in principle. But the detector is trained using annotations which are not precise enough to do good speech activity detection. The CHiME-6 recipe trains a SAD on 16k data, and the model is also available on kaldi-asr.org.

RuslanSel commented:
I've got an error message,
when I ran stage 8 local/train_overlap_detector.sh.
steps/nnet3/get_egs_targets.sh: number of utterances 23 in your training data is too small versus --num-utts-subset=12
... you probably have so little data that it doesn't make sense to train a neural net.

desh2608 commented:
@RuslanSel can you check if your training data was prepared correctly?

RuslanSel commented:
My data/train/wav.scp has 23 Mix-Headset.wav files.
Is this the right amount?

desh2608 commented:
That doesn't sound right. I have 133 Mix-Headset wav files in my training set. You can still train with that data, by setting `--num-utts-subset` to less than 0.25*23, i.e., 5 or fewer, in `train_tdnn_lstm_1a.sh`, but I doubt the neural network would be able to learn much with so little data.

RuslanSel commented:
Thanks.
Something wrong with my download.
What $mic is right for 
local/ami_download.sh $mic $AMI_DIR
cause now mic: unbound variable

desh2608 commented:
You can set `mic=ihm` for this recipe.

johnjosephmorgan commented:
> I have 132 files in data/tran_ovl/wav.scp.
Maybe I'm missing 1 file?
@desh2608 could you verify that you have 133 files and not 132.

> On Oct 5, 2020, at 2:48 PM, RuslanSel <notifications@github.com> wrote:
> 
> 
> Thanks.
> Something wrong with my download.
> What $mic is right for
> local/ami_download.sh $mic $AMI_DIR
> cause now mic: unbound variable
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/kaldi-asr/kaldi/pull/4265#issuecomment-703820158>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJL7AHWT427CRFUDIILPQDSJIIJRANCNFSM4RHZTEUQ>.
>

RuslanSel commented:
I set mic=ihm, but there are zero Mixed-Headset.wavs after download.
Only Headset files.

johnjosephmorgan commented:
Hello @desh2608  Could you post the trained overlap models to the kaldi webpage?
